### PR TITLE
Improved and refactored text detection sample in dnn module

### PR DESCRIPTION
**Pull request by gursimarsingh**
Clean up samples: #25006 

This pull requests merges and simplifies different text detection samples in dnn module of opencv in to one file. An option has been provided to choose the detection model from EAST or DB

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/text_detection.cpp [hunk 1]
```diff
@@ -20,65 +20,104 @@
 #include <iostream>
 #include <fstream>
 
+// Including OpenCV core, image processing, and deep neural network modules
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:23]: Not necessary comment

File: samples/dnn/text_detection.cpp [hunk 2]
```diff
@@ -20,65 +20,104 @@
...
+    Mat frame = imread(samples::findFile(parser.get<String>("input")));
+
+    // Initializing and configuring the text detection model based on the provided config
+    if (detModelConfig == "east") {
+        // EAST Detector initialization
+        TextDetectionModel_EAST detector(detModelPath);
+        detector.setConfidenceThreshold(confThreshold)
+                .setNMSThreshold(nmsThreshold);
+        // Setting input parameters specific to EAST model
+        detector.setInputParams(1.0, Size(width, height), Scalar(123.68, 116.78, 103.94), true);
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:99]: The only generic parameter is `Size(width, height)`. Please update `TextDetectionModel_EAST_Impl` constructor with the rest of preprocessing options so in sample you just need to call `detector.setInputSize()`

 https://github.com/opencv/opencv/blob/f2c3d4dfe3843898d969a45e43d981e8af967bd5/modules/dnn/src/model.cpp#L1115

File: samples/dnn/text_detection.cpp [hunk 3]
```diff
@@ -20,65 +20,104 @@
...
+    }
+    else if (detModelConfig == "db") {
+        // DB Detector initialization
+        TextDetectionModel_DB detector(detModelPath);
+        detector.setBinaryThreshold(binThresh)
+                .setPolygonThreshold(polyThresh)
+                .setUnclipRatio(unclipRatio)
+                .setMaxCandidates(maxCandidates);
+        // Setting input parameters specific to DB model
+        detector.setInputParams(1.0 / 255.0, Size(width, height), Scalar(122.67891434, 116.66876762, 104.00698793));
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:111]: Same request as for EAST. Let's move magic constants to `TextDetectionModel_DB` constructor

File: samples/dnn/text_detection.cpp [hunk 4]
```diff
@@ -88,90 +127,76 @@ int main(int argc, char** argv)
...
-    Mat frame;
-    while (waitKey(1) < 0)
-    {
-        cap >> frame;
-        if (frame.empty())
-        {
-            waitKey();
-            break;
+    // Process detected text regions for recognition
+    Mat frame2 = frame.clone();
```
Comment (dkurt) [inline samples/dnn/text_detection.cpp:144]: Why `frame2` is required?

File: samples/dnn/text_detection.cpp (hunk 2 above)
Comment (abhishekGola54) [reply samples/dnn/text_detection.cpp:99]: @dkurt The other parameters can be generic if someone trains EAST model on custom dataset. We can add the default values in the constructor, but I propose that the call should remain setInputParams in the sample to show its usage

**Comment by asmorkalov** [comment]
@gursimarsingh I formatted sample a bit.

**Comment by asmorkalov** [comment]
Several notes on PR:
- TextDetectionModel section of doc/tutorials/dnn/dnn_text_spotting/dnn_text_spotting.markdown does not correspond to the sample. The command line parameters does not match.
- I propose to add `models.yml` support to the sample to use `download_models.py`, and use pre-processing parameters from it. See samples/dnn/object_detection.cpp as example.
- As alternative, we need command line parameters example or / and set default parameters to use samples without parameters.
-

**Comment by gursimarsingh** [comment]
@asmorkalov Can you check now?

File: samples/dnn/text_detection.py [hunk 5]
```diff
@@ -60,179 +68,88 @@ def fourPointsTransform(frame, vertices):
...
-
-        # Put efficiency information
-        label = 'Inference time: %.2f ms' % (tickmeter.getTimeMilli())
-        cv.putText(frame, label, (0, 15), cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0))
-
-        # Display the frame
-        cv.imshow(kWinName, frame)
-        tickmeter.reset()
+    args = get_args_parser()
+    args.model = findFile(args.model)
```
Comment (asmorkalov) [inline samples/dnn/text_detection.py:77]: There is no model option in command line params

File: samples/dnn/text_detection.py [hunk 6]
```diff
@@ -60,179 +68,88 @@ def fourPointsTransform(frame, vertices):
...
+    args.input = findFile(args.input)
+    args.vocabulary_path = findFile(args.vocabulary_path)
+
+    frame = cv2.imread(args.input)
+
+    if(args.alias == "DB"):
```
Comment (asmorkalov) [inline samples/dnn/text_detection.py:83]: There is no alias option in command line params

File: samples/dnn/text_detection.py (hunk 5 above)
Comment (gursimarsingh) [reply samples/dnn/text_detection.py:77]: @asmorkalov The model option is added using the models.yml via args.zoo

```python
    args, _ = parser.parse_known_args()
    add_preproc_args(args.zoo, parser, 'text_detection')
```

File: samples/dnn/text_detection.py (hunk 6 above)
Comment (gursimarsingh) [reply samples/dnn/text_detection.py:83]: @asmorkalov The alias option is added using the models.yml via args.zoo

```python
    args, _ = parser.parse_known_args()
    add_preproc_args(args.zoo, parser, 'text_detection')
```

Comment (gursimarsingh) [reply samples/dnn/text_detection.py:85]: We need to provide the detection model name and path to recognition model. I have updated the usage made the path to recognition model as required

`python text_detection.py DB --ocr=<path to recognition model>`

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
@gursimarsingh, the new visualization looks great!
@asmorkalov, can we merge this PR?

**Comment by asmorkalov** [comment]
@gursimarsingh please rebase and fix conflicts.

**Comment by asmorkalov** [comment]
There is some issue with OCR model search in Python example:
```
./bin/example_dnn_text_detection DB
[ WARN:0@0.006] global samples.cpp:61 findFile cv::samples::findFile('right.jpg') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/right.jpg'
[ WARN:0@1.698] global samples.cpp:61 findFile cv::samples::findFile('alphabet_36.txt') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/alphabet_36.txt'
0: 'ka'
1: 'santiram'
2: 'fundamentals'
3: 'it'
4: 'circuits'
5: 'awd'
6: 'devices'
7: 'electronics'
8: 'basic'
9: 'letusc'
10: 'elit'
alexander@paradox:~/Projects/OpenCV/opencv-next/samples/dnn$ python3 ./text_detection.py DB
[ WARN:0@0.021] global samples.cpp:61 findFile cv::samples::findFile('DB_IC15_resnet50.onnx') => ''
[ WARN:0@0.021] global samples.cpp:61 findFile cv::samples::findFile('right.jpg') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/right.jpg'
[ WARN:0@0.021] global samples.cpp:61 findFile cv::samples::findFile('alphabet_36.txt') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/alphabet_36.txt'
[ERROR] Please pass the path to the ocr model using --ocr to run the sample
```

**Comment by asmorkalov** [comment]
```
./bin/example_dnn_text_detection East
[ WARN:0@0.006] global samples.cpp:61 findFile cv::samples::findFile('right.jpg') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/right.jpg'
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(5.0.0-pre) /home/alexander/Projects/OpenCV/opencv-next/modules/dnn/src/tensorflow/tf_io.cpp:42: error: (-2:Unspecified error) FAILED: ReadProtoFromBinaryFile(param_file, param). Failed to parse GraphDef file: /home/alexander/Projects/OpenCV/opencv-next/samples/dnn/3ca8233d6edd748f7ed23246c8ca24cbf696bb94/frozen_east_text_detection.pb in function 'ReadTFNetParamsFromBinaryFileOrDie'

Core dumped
```

**Comment by gursimarsingh** [comment]
@asmorkalov The issue has been fixed and models.yaml has been updated. A print message has been added in download_model.yaml when trying to compute hashsum for a file which does not exist. Instead of hash mismatch error it now prints skipping hashsum check in that case

Python code has also been updated with visualisation same as cpp

**Comment by asmorkalov** [comment]
```
alexander@paradox:~/Projects/OpenCV/opencv-next/samples/dnn$ python3 download_models.py East
DownloadInstance <opencv_face_detector_fp16>
DownloadInstance <opencv_face_detector_uint8>
DownloadInstance <yolov8x>
DownloadInstance <yolov8s>
DownloadInstance <yolov8n>
DownloadInstance <yolov8m>
DownloadInstance <yolov8l>
DownloadInstance <yolov4>
DownloadInstance <yolov4-tiny>
DownloadInstance <yolov3>
DownloadInstance <tiny-yolo-voc>
DownloadInstance <yolov8>
DownloadInstance <ssd_caffe>
DownloadInstance <ssd_tf>
DownloadInstance <faster_rcnn_tf>
DownloadInstance <squeezenet>
DownloadInstance <googlenet>
DownloadInstance <resnet>
DownloadInstance <fcnresnet50>
DownloadInstance <fcnresnet101>
DownloadInstance <u2netp>
DownloadInstance <DB>
DownloadInstance <East>
Model: East
  Working on East
  Getting file frozen_east_text_detection.pb
  expected SHA1: fffabf5ac36f37bddf68e34e84b45f5c4247ed06
  there is already a file with the same name
  actual SHA1:fffabf5ac36f37bddf68e34e84b45f5c4247ed06
  hash match - file already exists, skipping
DownloadInstance <OCR>
DownloadInstance <dexined>
alexander@paradox:~/Projects/OpenCV/opencv-next/samples/dnn$ 
alexander@paradox:~/Projects/OpenCV/opencv-next-build$ ./bin/example_dnn_text_detection East
[ WARN:0@0.006] global samples.cpp:61 findFile cv::samples::findFile('right.jpg') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/right.jpg'
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(5.0.0-pre) /home/alexander/Projects/OpenCV/opencv-next/modules/dnn/src/tensorflow/tf_io.cpp:42: error: (-2:Unspecified error) FAILED: ReadProtoFromBinaryFile(param_file, param). Failed to parse GraphDef file: /home/alexander/Projects/OpenCV/opencv-next/samples/dnn/3ca8233d6edd748f7ed23246c8ca24cbf696bb94/frozen_east_text_detection.pb in function 'ReadTFNetParamsFromBinaryFileOrDie'

Core dumped
alexander@paradox:~/Projects/OpenCV/opencv-next-build$ ls -la /home/alexander/Projects/OpenCV/opencv-next/samples/dnn/3ca8233d6edd748f7ed23246c8ca24cbf696bb94/frozen_east_text_detection.pb
-rw-rw-r-- 1 alexander alexander 89824490 сен  6 15:34 /home/alexander/Projects/OpenCV/opencv-next/samples/dnn/3ca8233d6edd748f7ed23246c8ca24cbf696bb94/frozen_east_text_detection.pb
```

**Comment by asmorkalov** [comment]
DB model works correctly with C++ sample

**Comment by asmorkalov** [comment]
```
 python3 ./text_detection.py DB
[ WARN:0@0.023] global samples.cpp:61 findFile cv::samples::findFile('DB_IC15_resnet50.onnx') => ''
[ WARN:0@0.023] global samples.cpp:61 findFile cv::samples::findFile('VGG_CTC.onnx') => ''
[ WARN:0@0.023] global samples.cpp:61 findFile cv::samples::findFile('right.jpg') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/right.jpg'
[ WARN:0@0.023] global samples.cpp:61 findFile cv::samples::findFile('alphabet_36.txt') => '/home/alexander/Projects/OpenCV/opencv-next/samples/data/alphabet_36.txt'
0: 'ka'
1: 'santiram'
2: 'fundamentals'
3: 'it'
4: 'circuits'
5: 'awd'
6: 'devices'
7: 'electronics'
8: 'basic'
9: 'letusc'
10: 'elit'
Traceback (most recent call last):
  File "./text_detection.py", line 197, in <module>
    main()
  File "./text_detection.py", line 191, in main
    stacked = np.hstack([frame, board],dtype = "uint8")
  File "<__array_function__ internals>", line 4, in hstack
TypeError: _vhstack_dispatcher() got an unexpected keyword argument 'dtype'
```

**Comment by gursimarsingh** [comment]
> /home/alexander/Projects/OpenCV/opencv-next/samples/dnn/3ca8233d6edd748f7ed23246c8ca24cbf696bb94/frozen_east_text_detection.pb

@asmorkalov Delete the old model file and please download again using download_models.py.

**Comment by gursimarsingh** [comment]
> TypeError: _vhstack_dispatcher() got an unexpected keyword argument 'dtype'

@asmorkalov Seems like a numpy version issue. Will change this to cv2.hconcat()

**Comment by asmorkalov** [comment]
@fengyuentau Could you take a look on East model crash?

**Comment by gursimarsingh** [comment]
@asmorkalov The issue is that the tar.gz file has been downloaded and renamed to .pb instead of extracting it. 

With the new download_models.py and models.yml it should work if you delete the old model file and it's folder

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**